Automattic/wp-calypso
# Domains: Prevent connection of domains that already point to WPCOM during sign-up

## Proposed Changes

- Updates the domain connection process during sign-up, preventing customers from adding a domain connection to a new site if they would need to prove that they own the domain (pcYYhz-kh-p2). A message asking users to purchase a plan to proceed is shown:

<img width="1020" alt="Screenshot 2023-06-06 at 12 47 51" src="https://github.com/Automattic/wp-calypso/assets/5324818/2d5afeff-c69d-4399-810b-ade7867ce4b6">

- The same should happen when trying to connect a domain to am existing free site (i.e. a site without a paid plan). A message asking users to purchase a plan to proceed should be shown:

<img width="1084" alt="Screenshot 2023-06-06 at 12 50 01" src="https://github.com/Automattic/wp-calypso/assets/5324818/19081fd9-8e15-4ea7-a241-7a4405bbc47a">

### Context

Before D112394-code was deployed on 2023-05-31, we weren't checking domain ownership for any domain connections during sign-up, only for existing sites with plans. This means that a bad actor could create a new site with a plan and a domain connection to a domain or subdomain that was previously mapped to us and redirect that domain/subdomain to another, possibly malicious, site.

This PR fixes that problem by requiring users to purchase a plan if the domain connection they want to add requires verification. The connection can be verified and added after the user purchases a plan.

### Screen recordings

- Before:

https://github.com/Automattic/wp-calypso/assets/5324818/a00c9e22-d9f6-4867-bc5e-f4c3a4f09340

- After:

https://github.com/Automattic/wp-calypso/assets/5324818/a9174259-6c74-4ccf-87b9-2b909b53dd11

- After (existing free site without paid plan):

https://github.com/Automattic/wp-calypso/assets/5324818/db010ac7-0861-4094-89c6-c852c63ed8fe

## Testing Instructions

- In your backend, apply these changes 3084d-pb/#diff to disable caching and a11n bypassing
- Open the live Calypso link or build this branch locally
- Navigate to `/start`
- Select the "Use a domain I own" option
- Use a domain that already points to WPCOM (you can use `wpcom3.leotaba.com` if you want)
- Ensure you can't connect that domain during sign-up and a message asking you to buy a plan is shown
- Ensure the rest of sign-up continues as normal, with only a plan being purchased in the end

For an existing free site without a paid plan:
- Ensure you can't connect a domain that requires ownership verification and a message asking you to buy a plan is shown

Also important to test:
- Ensure you can connect a domain that isn't pointing to WPCOM during sign-up
- Try going back and forth along the steps of the sign-up flow and ensure nothing is broken
  - The back button, in particular, should go to different steps if you skipped the domain step or not
  - If you skipped the domain connection step, the plan step should return to the initial domains step
  - If you added a domain connection, the plan step should return to the "transfer or connect" step
- In an existing site, ensure you go through the ownership verification process if you try to connect a domain that points to WPCOM

## Pre-merge Checklist
